### ISSUE TITLE
UnicodeDecodeError: 'ascii' codec can't decode byte 0x83 in position 52

### ISSUE THREAD

**Issue by LexiconCode** [Danesprite/dragonfly#12]
This may be related to porting Dragonfly to Python 3.
It was first discovered when enabling Caster filter rules. It can be reproduced on initialization of Caster by adding in `_caster.py` and adding `from builtins import str` and `from __future__ import unicode_literals` to the top.

```
`Error loading _caster from C:\NatLink\NatLink\MacroSystem\_caster.py
Traceback (most recent call last):
  File "C:\NatLink\NatLink\MacroSystem\core\natlinkmain.py", line 322, in loadFile
    imp.load_module(modName,fndFile,fndName,fndDesc)
  File "C:\NatLink\NatLink\MacroSystem\_caster.py", line 69, in <module>
    class MainRule(MergeRule):
  File "C:\NatLink\NatLink\MacroSystem\_caster.py", line 131, in MainRule
    "disable": False
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\grammar\elements_compound.py", line 323, in __init__
    child = Compound(spec=k, value=v, extras=extras)
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\grammar\elements_compound.py", line 270, in __init__
    element = self._parser.parse(spec)
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 52, in parse
    for result in generator:
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 413, in parse
    try: path[-1].next()
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 472, in parse
    try: path[-1].next()
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 413, in parse
    try: path[-1].next()
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 547, in parse
    for result in child.parse(state):
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 413, in parse
    try: path[-1].next()
  File "C:\Python27\lib\site-packages\dragonfly-0.6.6b1-py2.7.egg\dragonfly\parser.py", line 708, in parse
    while not state.finished() and state.peek(1) in self._set:
UnicodeDecodeError: 'ascii' codec can't decode byte 0x83 in position 52: ordinal not in range(128)`
```
Full issue is posted https://github.com/synkarius/caster/issues/236

Let me know how I can help!

**Comment by Danesprite** [comment]
Thanks for filing this. I remember you mentioned something similar in the gitter channel a while ago concerning window titles. I'll do some testing and see what I can do. It may be a good idea to add a test method in `dragonfly/test/test_parser.py` for testing this, it only covers ascii at the moment.

It may be an issue with Python 3 porting too I guess. There is a lot of encoding/decoding between the `ascii`, `windows-1252` and `utf-8` encodings in dragonfly that makes porting to Python 3 complicated.

**Comment by Danesprite** [comment]
Okay I've made a few changes which seem to stop a `UnicodeDecodeError` being raised. Adding `from __future__ import unicode_literals` to the top of `_caster.py` also seemed to fix it for me. Oddly, the modified parser tests pass either way. Let me know if it works for you.

**Comment by LexiconCode** [comment]
Excellent, I will take a look! Thank you for looking into it!

>  Adding `from __future__ import unicode_literals` to the top of `_caster.py` 

That is indeed strange. Is this before or after implementing your fix and are you testing and the caster development branch (shouldn't make a difference but...)? 

The imports look like the following?
```
from __future__ import unicode_literals
import logging
logging.basicConfig()

import time
from dragonfly import (Key, Function, Grammar, Playback, Dictation, Choice, Pause)
from caster.lib.ccr.standard import SymbolSpecs
```
I will test your fix today or tomorrow then let you know!

**Comment by Danesprite** [comment]
You're welcome!

Never mind actually, I just tested again without the fix and no, the `__future__` import doesn't fix it. I must have tested with the fix accidentally that time.

**Comment by LexiconCode** [comment]
Sorry my computer went nuclear and blew up the MBR. Which I'm trying to pain stakingly rebuild the save my windows install. I don't want to lose the one of the activations for DNS. As soon as the PC is functional I'll get back to this.

**Comment by Danesprite** [comment]
Oh no! I'm very sorry to hear that :cry:
Don't worry about it, no rush.

**Comment by Versatilus** [comment]
One place this happens for me is where Dragon tries to process (for `AppContext`) the title of certain windows that have non-ASCII characters in them such as the – (emdash) in the title of Atom windows because the default codec is `ASCII`. I haven't looked at it since to try debugging further.

**Comment by Danesprite** [comment]
Yep, that is a related issue. The parser needs to be updated to use `utf-8` instead of `ascii`. I'll be looking into this some more soon.

**Comment by Danesprite** [comment]
I've just verified that 9fe1d04 did fix this. The related window title bug is probably fixed too, if not by that commit, then by 253dd15203c7f382943f4f97b27ad77e6fce6f43 when it gets merged to master.

**Comment by LexiconCode** [comment]
Excellent I will take a look and pass on the awesome news!
